File: pc/src/host/BridgeEndpointPolicy.test.ts
import assert from "node:assert/strict";
import test from "node:test";
import { normalizeBridgeEndpoint } from "./BridgeEndpointPolicy.js";

test("bridge endpoint policy requires TLS for ordinary network hosts", () => {
  assert.equal(normalizeBridgeEndpoint("WSS://Bridge.Example:8788").url, "wss://bridge.example:8788/phone");
  assert.throws(() => normalizeBridgeEndpoint("ws://192.168.1.20:8788/phone"), /require wss/);
  assert.throws(() => normalizeBridgeEndpoint("ws://bridge.example:8788/phone"), /require wss/);
});

test("bridge endpoint policy keeps bounded development exceptions", () => {
  assert.equal(normalizeBridgeEndpoint("ws://127.0.0.1:8788").security, "local-development");
  assert.equal(normalizeBridgeEndpoint("ws://[::1]:8788/phone").security, "local-development");
  assert.throws(() => normalizeBridgeEndpoint("ws://100.88.12.34:8788/phone"), /explicitly trusted/);
  assert.equal(normalizeBridgeEndpoint("ws://100.88.12.34:8788/phone", {
    allowInsecureTrustedOverlay: true
  }).security, "trusted-overlay-development");
  assert.equal(normalizeBridgeEndpoint("ws://host.tailnet.ts.net:8788/phone", {
    allowInsecureTrustedOverlay: true
  }).security, "trusted-overlay-development");
  assert.throws(() => normalizeBridgeEndpoint("ws://192.168.1.20:8788/phone", {
    allowInsecureTrustedOverlay: true
  }), /require wss/);
});

File: android/app/src/test/java/dev/androidagent/BridgeEndpointPolicyTest.kt
package dev.androidagent

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test

class BridgeEndpointPolicyTest {
    @Test
    fun requiresTlsForOrdinaryNetworkEndpoints() {
        assertEquals(
            "wss://bridge.example:8788/phone",
            BridgeEndpointPolicy.normalize("WSS://Bridge.Example:8788")?.url
        )
        assertNull(BridgeEndpointPolicy.normalize("ws://192.168.1.20:8788/phone"))
        assertNull(BridgeEndpointPolicy.normalize("ws://bridge.example:8788/phone"))
    }

    @Test
    fun permitsOnlyBoundedCleartextDevelopmentEndpoints() {
        assertEquals(
            BridgeEndpointSecurity.LocalDevelopment,
            BridgeEndpointPolicy.normalize("ws://127.0.0.1:8788")?.security
        )
        assertNull(BridgeEndpointPolicy.normalize("ws://100.88.12.34:8788/phone"))
        assertEquals(
            BridgeEndpointSecurity.TrustedOverlayDevelopment,
            BridgeEndpointPolicy.normalize("ws://100.88.12.34:8788/phone", allowInsecureTrustedOverlay = true)?.security
        )
        assertNull(BridgeEndpointPolicy.normalize("ws://192.168.1.20:8788/phone", allowInsecureTrustedOverlay = true))
    }
}
